### PR TITLE
Address remaining comments from #30

### PR DESCRIPTION
**Pull request by alexfikl**
This just adds docs and comments and some types here and there following the review from #30.

There is one actual change: `L` and `R` are changed from `(nclusters, ncluster)` object arrays to just `(nclusters,)` since they are block diagonal anyways. Not sure why I made them 2D arrays in the first place..

### CODE REVIEW

File: pytential/linalg/skeletonization.py [hunk 1]
```diff
@@ -400,44 +480,39 @@ def __getitem__(self, i: int) -> Tuple[np.ndarray, np.ndarray]:
         return [pxymat_i, nbrmat_i]
```
Comment (inducer) [inline pytential/linalg/skeletonization.py:524]: I thought I was promised a tuple! :)

File: pytential/linalg/utils.py [hunk 2]
```diff
@@ -353,6 +353,30 @@ def cluster_skeletonization_error(
...
+        E^S_{ij} = \|A_{ij} - S_{ij} R_j\|
+
+    where :math:`A_{ij}` is the exact interaction between the two clusters,
+    :math:`(L_i, R_j)` are the ID interpolation matrices and
+    :math:`(T_{ij}, S_{ij})` are the target and source skeleton matrices.
+    The exact matrix and the skeleton matrices are extracted from the full
+    matrix *mat*.
+
+    :arg ord: the type of the matrix norm used to compute errors, as described
+        in :func:`numpy.linalg.norm`.
```
Comment (inducer) [inline pytential/linalg/utils.py:376]: State that this identifies the norm $\|\cdot\|$ used above.

File: pytential/linalg/utils.py [hunk 3]
```diff
@@ -402,6 +426,23 @@ def skeletonization_error(
         mat: np.ndarray, skeleton: "SkeletonizationResult", *,
         ord: Optional[float] = None,
         relative: bool = False) -> np.ndarray:
+    r"""Computes the skeletonization error for the entire matrix *mat*.
+
+    The error computed here is given by
+
+    .. math::
+
+        E = \|A - L S R\|,
```
Comment (inducer) [inline pytential/linalg/utils.py:437]: Specify the norm being used, as above.

File: test/test_linalg_skeletonization.py [hunk 4]
```diff
@@ -341,9 +343,20 @@ def run_skeletonize_by_proxy(actx, case, resolution,
...
+    # SKELETONIZE_TEST_CASES[0], SKELETONIZE_TEST_CASES[1],
+    SKELETONIZE_TEST_CASES[2],
+    ])
 def test_skeletonize_by_proxy(actx_factory, case, visualize=False):
-    """Test single-level level skeletonization accuracy."""
+    r"""Test single-level level skeletonization accuracy. Checks that the error
+    satisfies :math:`e < c \epsilon_{id}` for a fixed ID tolerance and an
+    empirically determined (not too huge) :math:`c`.
+
+    NOTE: this test is mostly
```
Comment (inducer) [inline test/test_linalg_skeletonization.py:352]: This looks

File: test/test_linalg_skeletonization.py [hunk 5]
```diff
@@ -382,7 +400,10 @@ def test_skeletonize_by_proxy(actx_factory, case, visualize=False):
 def test_skeletonize_by_proxy_convergence(
         actx_factory, case, weighted=True,
         visualize=False):
-    """Test single-level level skeletonization accuracy."""
+    r"""Test single-level level skeletonization accuracy. Checks that the
```
Comment (inducer) [inline test/test_linalg_skeletonization.py:398]: Extra "level"?

File: pytential/linalg/skeletonization.py [hunk 6]
```diff
@@ -52,6 +52,50 @@
 
 # {{{ wrangler
 
+# Adding weights to proxy evaluation: Why? How? Huh?
```
Comment (inducer) [inline pytential/linalg/skeletonization.py:55]: :rofl:

File: pytential/linalg/skeletonization.py [hunk 7]
```diff
@@ -52,6 +52,50 @@
...
+# proxy-based skeletonization, namely
+# 1. with the proxies + neighbors when the current cluster is the source
+# 2. with the proxies + neighbors when the current cluster is the target
+#
+# The operator that we want to skeletonize at the end of the day is the full
+# layer potential, that contains quadrature weights and area elements. That
+# means that it is in our interest to include some of that information in
+# the proxy + neighbor interaction matrices that get passed to the ID
+# (interpolative decomposition), so that it has a better chance at guessing a
+# good range!
```
Comment (inducer) [inline pytential/linalg/skeletonization.py:68]: Maybe rephrase this to emphasize weighting of relative error?

**Review by inducer — APPROVED** [review]
Thanks for following up! This actually makes a great big difference in how easy to follow I find that code. A few things below, good to go once those are addressed.

File: pytential/linalg/skeletonization.py (hunk 1 above)
Comment (alexfikl) [reply pytential/linalg/skeletonization.py]: Haha, indeed! Fixed in https://github.com/inducer/pytential/compare/440161e74f559e51120245ce59f8c1a6c4f15d77..0617bdafd0a7a37c2e162b752371ab908e456ad5.

File: pytential/linalg/utils.py (hunk 2 above)
Comment (alexfikl) [reply pytential/linalg/utils.py]: Added a sentence in https://github.com/inducer/pytential/compare/440161e74f559e51120245ce59f8c1a6c4f15d77..0617bdafd0a7a37c2e162b752371ab908e456ad5.

File: pytential/linalg/utils.py (hunk 3 above)
Comment (alexfikl) [reply pytential/linalg/utils.py:438]: https://github.com/inducer/pytential/compare/440161e74f559e51120245ce59f8c1a6c4f15d77..0617bdafd0a7a37c2e162b752371ab908e456ad5

File: test/test_linalg_skeletonization.py (hunk 4 above)
Comment (alexfikl) [reply test/test_linalg_skeletonization.py]: Fixed in https://github.com/inducer/pytential/compare/440161e74f559e51120245ce59f8c1a6c4f15d77..0617bdafd0a7a37c2e162b752371ab908e456ad5!

File: test/test_linalg_skeletonization.py (hunk 5 above)
Comment (alexfikl) [reply test/test_linalg_skeletonization.py]: And copy-pasted below :facepalm: Fixed in https://github.com/inducer/pytential/compare/440161e74f559e51120245ce59f8c1a6c4f15d77..0617bdafd0a7a37c2e162b752371ab908e456ad5

File: pytential/linalg/skeletonization.py (hunk 7 above)
Comment (alexfikl) [reply pytential/linalg/skeletonization.py:68]: Not quite sure what you had in mind here?

Comment (inducer) [reply pytential/linalg/skeletonization.py:68]: I feel that the intutition we had built was that the ID will ultimately seek to squeeze the matrix norm of the error below the supplied bound, i.e. a relative error measure. That's easier for columns with small norms than those with big norms, and so it'll "make more effort" to make big-norm columns "happy" than others. But for our application, the unequal weighting is spurious, that's why we try to fix it.

I was hoping the comment narrative could reflect that train of thought.

Comment (alexfikl) [reply pytential/linalg/skeletonization.py]: Ah, right! That totally slipped my mind. I added something like that in 
https://github.com/inducer/pytential/compare/0617bdafd0a7a37c2e162b752371ab908e456ad5..2e897125c3e199988600d73c65ca28c1a7d55c1c
(let me know if it's not clear enough!)

Comment (inducer) [reply pytential/linalg/skeletonization.py]: That works, thanks!
